conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanBuilderGenerator.java

```diff
@@ -673,17 +658,29 @@ private MethodSpec createCollectionSetter(String prefix, EnrichedField enriched,
                 .build();
     }
 
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
     private CodeBlock typeAwareAssignment(EnrichedField enriched, Type type, boolean shouldClearFirst) {
         FieldSpec spec = enriched.poetSpec();
         if (type.accept(TypeVisitor.IS_LIST) || type.accept(TypeVisitor.IS_SET)) {
             if (shouldClearFirst) {
-                return CodeBlocks.statement(
-                        "this.$1N = $2T.new$3T($4L)",
-                        spec.name,
-                        ConjureCollections.class,
-                        type.accept(COLLECTION_CONCRETE_TYPE),
-                        Expressions.requireNonNull(
-                                spec.name, enriched.fieldName().get() + " cannot be null"));
+                CollectionType collectionType = getCollectionType(type);
+                if (collectionType.useNonNullFactory()) {
+                    return CodeBlocks.statement(
+                            "this.$1N = $2T.newNonNull$3T($4L)",
```

Comment:
prefixing with `newNonNull` and adding the null checks to ConjureCollections seems like the simplest way to do this apart from creating a NonNullArrayList class.